File: Algorithm/KF-classify.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

void kalman_filter_classify(int measurement) {
    static double x = 0.0; // 初始位置
    static double E_est = 0.5; // 初始误差协方差
    const double Q = 0.003; // 过程噪声协方差
    const double E_mea = 0.35; // 观测噪声协方差
    // 状态转移矩阵
    const double F = 1.0; // 假设位置不变
    // 观测矩阵
    const double H = 1.0;

    // 预测步骤
    x = F * x;
    E_est = F * E_est * F + Q;

    double y = measurement - H * x;
    double S = H * E_est * H + E_mea;

    double K = E_est * H / S; // 计算卡尔曼增益
    x = x + K * y; // 更新状态估计
    E_est = (1 - K * H) * E_est; // 更新误差协方差

    // 分类
    int classification = x >= 0.5 ? 1 : 0;
    printf("Measure = %d, Predict = %.5f, E_est = %.10f, K = %.5f, Classification = %d\n", 
           measurement, x, E_est, K, classification);
}

int main() {
    srand(time(0)); // 初始化随机数生成器

    int measurements[2010];
    for (int i = 0; i < 1000; i++) {
        measurements[i] = rand() % 2;
    }
    for (int i = 1000; i < 2000; i++) {
        measurements[i] = 1;
    }
    for (int i = 2000; i < 2010; i++) {
        measurements[i] = 0;
    }

    for (int i = 0; i < 2010; i++) {
        kalman_filter_classify(measurements[i]);
    }

    return 0;
}

File: Algorithm/KF_classify.py
import numpy as np
import random
import matplotlib.pyplot as plt

class KalmanFilter:
    def __init__(self, std_acc, x_std_meas, y_std_meas):
        # 状态向量 [x, y]
        self.x = np.matrix([[0], [0]])
        
        # 过程噪声协方差
        self.Q = np.matrix([[std_acc**2, 0],
                            [0, std_acc**2]])
        
        # 测量误差协方差
        self.R = np.matrix([[x_std_meas**2, 0],
                            [0, y_std_meas**2]])
        
        # 状态转移矩阵
        self.F = np.matrix([[1, 0],
                            [0, 1]])
        
        # 观测矩阵
        self.H = np.matrix([[1, 0],
                            [0, 1]])
        # 估计误差协方差矩阵
        self.P = 10*np.eye(self.F.shape[1]) 
    
    def predict(self):
        # 状态预测
        self.x = self.F @ self.x

        # 误差协方差预测
        self.P = self.F @ self.P @ self.F.T + self.Q

        return self.x

    def update(self, z):
        # 计算卡尔曼增益
        S = self.H @ self.P @ self.H.T + self.R
        K = self.P @ self.H.T @ np.linalg.inv(S)
        #print('K=',K[0][0])
        # 更新状态向量
        y = z - self.H @ self.x
        self.x = self.x + K @ y

        # 更新误差协方差矩阵
        I = np.eye(self.H.shape[1])
        self.P = (I - K @ self.H) @ self.P
        #print(f"更新后的误差协方差矩阵P: {self.P[0][0]}\n")

        return self.x

if __name__ == "__main__":
    # 示例使用
    std_acc = 1  # 过程噪声标准差
    x_std_meas = 5  # x测量噪声标准差
    y_std_meas = 5  # y测量噪声标准差

    kf = KalmanFilter(std_acc, x_std_meas, y_std_meas)

    # # 示例使用
    trajectory = []
    x = 0
    y = 0
    for _ in range(500):
        x += random.uniform(-1, 1.1)
        y += random.uniform(-1, 1)
        trajectory.append([x, y])

    measurements = []
    predictions = []

    for measurement in trajectory:
        z = np.matrix(measurement).T
        kf.predict()
        prediction = kf.update(z)
        measurements.append(measurement)
        predictions.append(prediction.T.tolist()[0])
        #print(f"measure: {measurement}, predict: {prediction.T.tolist()[0]}")

    # 转换为numpy数组以便绘图
    measurements = np.array(measurements)
    predictions = np.array(predictions)

    # 绘制测量值和预测值的曲线
    plt.figure(figsize=(10, 5))
    plt.plot(measurements[:, 0], measurements[:, 1], label='Measure', marker='o')
    plt.plot(predictions[:, 0], predictions[:, 1], label='Predict', marker='x')
    plt.xlabel('X ')
    plt.ylabel('Y ')
    plt.title('KF_Predict')
    plt.legend()
    plt.grid()
    plt.show()
